app/utility/base_knowledge_svc.py

```diff
@@ -36,89 +36,90 @@ def _add_fact(self, fact, constraints=None):
             for k, v in updates.items():
                 if getattr(match, k, False):
                     setattr(match, k, v)
 
-    def _get_facts(self, criteria, restrictions=None):
+    async def _get_facts(self, criteria, restrictions=None):
         """
         Retrieve a fact from the internal store
         :param criteria: dictionary containing fields to match on
         :return: list of facts matching the criteria
         """
-        complete_list = self._locate('facts', query=criteria)
-        return self._apply_restrictions(complete_list, restrictions)
+        complete_list = await self._locate('facts', query=criteria)
+        return await self._apply_restrictions(complete_list, restrictions)
 
-    def _delete_fact(self, criteria):
+    async def _delete_fact(self, criteria):
         """
         Delete a fact from the internal store
         :param criteria: dictionary containing fields to match on
         """
-        return self._remove('facts', criteria)
+        return await self._remove('facts', criteria)
 
-    def _get_meta_facts(self, meta_fact=None, agent=None, group=None):
+    async def _get_meta_facts(self, meta_fact=None, agent=None, group=None):
         # Returns the complete set of facts associated with a meta-fact construct
         raise NotImplementedError
 
-    def _get_fact_origin(self, fact):
+    async def _get_fact_origin(self, fact):
         # Retrieve the specific origin of a fact. If it was learned in the current operation, parse through links to
         # identify the host it was discovered on.
         raise NotImplementedError
 
     # -- Relationships API --
 
-    def _get_relationships(self, criteria, restrictions=None):
+    async def _get_relationships(self, criteria, restrictions=None):
         """
         Retrieve relationships from the internal store
         :param criteria: dictionary containing fields to match on
         :return: list of matching relationships
         """
-        complete_list = self._locate('relationships', query=criteria)
-        return self._apply_restrictions(complete_list, restrictions)
+        complete_list = await self._locate('relationships', query=criteria)
+        return await self._apply_restrictions(complete_list, restrictions)
 
-    def _add_relationship(self, relationship, constraints=None):
+    async def _add_relationship(self, relationship, constraints=None):
         """
         Add a relationship to the internal store
         :param relationship: Relationship object to add
         :param constraints: optional constraints on the use of the relationship
         """
         if not any((x.source == relationship.source) and (x.edge == relationship.edge) and
-                   (x.target == relationship.target) for x in self.fact_ram['relationships']):
+                   (x.target == relationship.target) and (x.origin == relationship.origin or not relationship.origin)
+                   for x in self.fact_ram['relationships']):
```

Comment:
Probably a future thing, but having a way to lookup objects without having to enumerate through all of them would be nice for performance in large operations with lots of relationships/facts in the knowledge store.

Reply:
Yep, but I'd put that specifically in the 'better backend' space - there's a bunch of fancy things you can do with other data store types to make that really fast and efficient, much more than we could get in a pure python implementation without re-implementing those store types.

Reply:
👍 deal